test/cpp/tensorexpr/test_cuda.cpp

```diff
@@ -2344,6 +2344,315 @@ TEST(Cuda, MaskMultiDimMultiLevel_CUDA) {
   cudaFree(d_dev);
 }
 
+TEST(Cuda, DependencyAcrossLoops_CUDA) {
+  // for (int x = 0; x < 10; x++) {
+  //   b[x] = a[x];
+  // }
+  // for (int y = 0; y < 10; y++) {
+  //   c[y] = b[y+1];
+  // }
+  KernelScope kernel_scope;
+
+  Placeholder a("a", kInt, {50});
+  Tensor* b =
+      Compute("b", {{50, "x"}}, [&](const VarHandle& i) { return a.load(i); });
+  Tensor* c = Compute(
+      "c", {{50, "y"}}, [&](const VarHandle& i) { return b->call(i + 1); });
+  LoopNest l({b, c});
+  std::vector<For*> loops = l.getLoopStmtsFor(b);
+  l.setGPUThreadIndex(loops[0], 0);
+  loops = l.getLoopStmtsFor(c);
+  l.setGPUThreadIndex(loops[0], 0);
+
+  l.prepareForCodegen();
+  Stmt* s = l.root_stmt();
+
+  CudaCodeGen cg(s, {a, b, c});
+  std::ostringstream oss;
+  oss << *cg.stmt();
+
+  const std::string& verification_pattern =
+      R"IR(
+        # CHECK: b[threadIdx.x] =
+        # CHECK: __syncthreads();
+        # CHECK: c[threadIdx.x] =
+      )IR";
+
+  torch::jit::testing::FileCheck().run(verification_pattern, oss.str());
+}
+
+TEST(Cuda, DependencyInALoop_CUDA) {
+  // for (int x = 0; x < 50; x++) {
+  //   A[x] = x;
+  //   B[x] = A[x+1];
```

Comment:
I've actually given this a bit more thoughts. In fact, I would argue that this loop nest is invalid from our IR perspective. 

The major reason is that the default loop semantics is "independent statement evaluation". It is the foundation that we can evaluate each index and its associated statement in parallel, or in arbitrary order. With your example, it has to be evaluated in the serial order. In fact, it will be illegal again if it is "B[x] = A[x - 1]". 

So if in your example: it is first transformed into:

`
for (int x = 0; x < 50; x++) {
    A[x] = x;
}
for (int x = 0; x < 50; x++) {
    B[x] = A[x + 1]
}
`

aka your dependency-across-loop example, that should be fine. 

In summary, I think if the codegen sees the LoopNest in this form, it should be an error. Before that, some transformations should have transformed the LoopNest into a valid form.